Fixes #35767 to allow for view creation into specified schema from [ Create a view ] in DBManager

gisn8 opened the pull request:
Fixes #35767 to allow for view creation into specified schema from [ Create a view ] in DBManager

## Description

Currently, using [ Create a view ] from DBManager for postgres automatically loads the view into the public schema named the entire string. It is reasonable for the user to assume that specifying the schema name as schema.view will create the view into the schema specified. Instead, it's creates a view called schema.view in the public schema. This is generally warned against.

Tested

In order to meet users' expectations, allowed for testing if the string has a "." in the name and to separate the string into two parts to feed into the quoteId which is already set up to accept schema and viewname as a list to process appropriately.

Recommended documentation:
"The Create a View button opens a dialog where you can enter the name for the new view. By default, the view is created in the public schema of the PostgreSQL database. However, you can specify a different schema by using the schema.viewname naming convention."

Backporting is recommended to meet user expectations.

Fixes #35767 to allow for view creation into specified schema from [ Create a view ] in DBManager

gisn8 committed:
Addressing Issue #35767 to allow for view creation into specified schema from [ Create a view ] in DBManager

nyalldawson commented on python/plugins/db_manager/db_plugins/postgis/connector.py:
```diff
@@ -1014,7 +1014,13 @@ def moveTable(self, table, new_table, new_schema=None):
 
     def createView(self, view, query):
-        sql = "CREATE VIEW %s AS %s" % (self.quoteId(view), query)
+        user_input = view
+        
+        if '.' in user_input: # To allow view creation into specified schema
+            schema, view_name = user_input.split('.')
+            sql = "CREATE VIEW %s AS %s" % (self.quoteId([schema, view_name]), query)
+        else: # No schema specified; uses public
+            sql = "CREATE VIEW %s AS %s" % (self.quoteId(view), query)
```
This should probably be hardened a little -- eg if the view name was entered badly with two periods, then we'll get a raw python exception here. How about:

```suggestion
        view_name_parts = view.split('.')
        if len(view_name_parts) == 2: # To allow view creation into specified schema
            schema, view_name = view_name_parts
            sql = "CREATE VIEW %s AS %s" % (self.quoteId([schema, view_name]), query)
        else: # No specific schema specified
            sql = "CREATE VIEW %s AS %s" % (self.quoteId(view), query)
```

gisn8 committed:
Handle invalid view names with more than one period

I've updated the code as recommended and added an error message should there there be too many periods. Thanks for the suggestion!

gisn8 committed:
Fix: clean up whitespace and formatting issues caught by Flake8

nyalldawson commented on python/plugins/db_manager/db_plugins/postgis/connector.py:
```diff
@@ -1014,7 +1014,16 @@ def moveTable(self, table, new_table, new_schema=None):
+
+        if len(view_name_parts) > 2:
+            # Raise an error when more than one period is used.
+            raise ValueError("Invalid view name: Please use the format 'schema.viewname', or enter only the viewname for the public schema.")
```
A ValueError won't get gracefully caught here -- this should probably be:

```suggestion
            raise DbError("Invalid view name: Please use the format 'schema.viewname', or enter only the viewname for the public schema.")
```

gisn8 committed:
Set raise error to DbError

Thanks! Much more user-friendly.

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>
